contracts/zero-ex/contracts/src/features/NativeOrdersFeature.sol

```diff
@@ -0,0 +1,984 @@
+                takerAmount: order.takerAmount,
+                takerTokenFillAmount: takerTokenFillAmount,
+                takerTokenFilledAmount: orderInfo.takerTokenFilledAmount
+            })
+        );
+
+        emit RfqOrderFilled(
+            orderInfo.orderHash,
+            order.maker,
+            taker,
+            address(order.makerToken),
+            address(order.takerToken),
+            results.takerTokenFilledAmount,
+            results.makerTokenFilledAmount,
+            results.ethProtocolFeePaid,
+            order.pool
+        );
+    }
+
+    /// @dev Settle the trade between an order's maker and taker.
+    /// @param settleInfo Information needed to execute the settlement.
+    /// @return takerTokenFilledAmount How much taker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _settleOrder(SettleOrderInfo memory settleInfo)
+        private
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        // Clamp the taker token fill amount to the fillable amount.
+        takerTokenFilledAmount = LibSafeMathV06.min128(
+            settleInfo.takerTokenFillAmount,
+            settleInfo.takerAmount.safeSub128(settleInfo.takerTokenFilledAmount)
+        );
+        // Compute the maker token amount.
+        // This should never overflow because the values are all clamped to
+        // (2^128-1).
+        makerTokenFilledAmount = uint128(
+            uint256(takerTokenFilledAmount)
+            * uint256(settleInfo.makerAmount)
+            / uint256(settleInfo.takerAmount)
+        );
+
+        if (takerTokenFilledAmount == 0 || makerTokenFilledAmount == 0) {
+            // Nothing to do.
+            return (0, 0);
+        }
+
+        // Update filled state for the order.
+        LibNativeOrdersStorage
+            .getStorage()
+            .orderHashToTakerTokenFilledAmount[settleInfo.orderHash] =
+            // OK to overwrite the whole word because we shouldn't get to this
+            // function if the order is cancelled.
+                settleInfo.takerTokenFilledAmount + takerTokenFilledAmount;
+
+        // Transfer taker -> maker.
+        LibTokenSpender.spendERC20Tokens(
+            settleInfo.takerToken,
+            settleInfo.taker,
+            settleInfo.maker,
+            takerTokenFilledAmount
+        );
+
+        // Transfer maker -> taker.
+        LibTokenSpender.spendERC20Tokens(
+            settleInfo.makerToken,
+            settleInfo.maker,
+            settleInfo.taker,
+            makerTokenFilledAmount
+        );
+    }
+
+    /// @dev Refund any leftover protocol fees in `msg.value` to `msg.sender`.
+    /// @param ethProtocolFeePaid How much ETH was paid in protocol fees.
+    function _refundExcessProtocolFeeToSender(uint256 ethProtocolFeePaid)
+        private
+    {
+        if (msg.value > ethProtocolFeePaid && msg.sender != address(this)) {
```

Comment:
> For small values it may be cheaper for the taker if we do not refund the protocol fee. I'd be down to add a lower bound that refundAmount must be exceed in order for us to refund.

I wasn't sure if we discussed this offline or if it just got lost in the mix. I do have a cheeky update though. Danning and I were looking into the data and it appears that there are several cases where folks are just 1 gwei off from the correct protocol fee. Which could be caused by a rounding error on their end, not sure. But that could be a min dust amount that we just don't refund. So like `if (ethProtocolFeePaid - msg.value <= 1 gwei) {do not refund}`. Wdyt?

Reply:
How often? Is this an arbitrary number we're going to be chasing forever? Is it more robust to say `1 * tx.gasprice`? I feel like if there's a rounding error, that's an error we should try to fix in the client before hacking a fix in the protocol. :shrug:

Reply:
*** Note to reader - this has been backlogged.